source/ptrace.man
PTRACE(2)                                                                                               Linux Programmer's Manual                                                                                               PTRACE(2)

NAME
       ptrace - process trace

SYNOPSIS
       #include <sys/ptrace.h>

       long ptrace(enum __ptrace_request request, pid_t pid,
                   void *addr, void *data);

DESCRIPTION
       The  ptrace() system call provides a means by which one process (the "tracer") may observe and control the execution of another process (the "tracee"), and examine and change the tracee's memory and registers.  It is primarily
       used to implement breakpoint debugging and system call tracing.

       A tracee first needs to be attached to the tracer.  Attachment and subsequent commands are per thread: in a multithreaded process, every thread can be individually attached to a (potentially  different)  tracer,  or  left  not
       attached and thus not debugged.  Therefore, "tracee" always means "(one) thread", never "a (possibly multithreaded) process".  Ptrace commands are always sent to a specific tracee using a call of the form

           ptrace(PTRACE_foo, pid, ...)

       where pid is the thread ID of the corresponding Linux thread.

       (Note that in this page, a "multithreaded process" means a thread group consisting of threads created using the clone(2) CLONE_THREAD flag.)

       A  process  can initiate a trace by calling fork(2) and having the resulting child do a PTRACE_TRACEME, followed (typically) by an execve(2).  Alternatively, one process may commence tracing another process using PTRACE_ATTACH
       or PTRACE_SEIZE.

       While being traced, the tracee will stop each time a signal is delivered, even if the signal is being ignored.  (An exception is SIGKILL, which has its usual effect.)  The tracer will be notified at its next call to waitpid(2)
       (or  one  of  the  related  "wait"  system calls); that call will return a status value containing information that indicates the cause of the stop in the tracee.  While the tracee is stopped, the tracer can use various ptrace
       requests to inspect and modify the tracee.  The tracer then causes the tracee to continue, optionally ignoring the delivered signal (or even delivering a different signal instead).

       If the PTRACE_O_TRACEEXEC option is not in effect, all successful calls to execve(2) by the traced process will cause it to be sent a SIGTRAP signal, giving the parent a chance to gain control before  the  new  program  begins
       execution.

       When the tracer is finished tracing, it can cause the tracee to continue executing in a normal, untraced mode via PTRACE_DETACH.

       The value of request determines the action to be performed:

       PTRACE_TRACEME
              Indicate that this process is to be traced by its parent.  A process probably shouldn't make this request if its parent isn't expecting to trace it.  (pid, addr, and data are ignored.)

              The  PTRACE_TRACEME  request is used only by the tracee; the remaining requests are used only by the tracer.  In the following requests, pid specifies the thread ID of the tracee to be acted on.  For requests other than
              PTRACE_ATTACH, PTRACE_SEIZE, PTRACE_INTERRUPT, and PTRACE_KILL, the tracee must be stopped.

       PTRACE_PEEKTEXT, PTRACE_PEEKDATA
              Read a word at the address addr in the tracee's memory, returning the word as the result of the ptrace() call.  Linux does not have separate text and data address spaces, so these two requests are currently  equivalent.
              (data is ignored; but see NOTES.)

       PTRACE_PEEKUSER
              Read  a word at offset addr in the tracee's USER area, which holds the registers and other information about the process (see <sys/user.h>).  The word is returned as the result of the ptrace() call.  Typically, the off‐
              set must be word-aligned, though this might vary by architecture.  See NOTES.  (data is ignored; but see NOTES.)

       PTRACE_POKETEXT, PTRACE_POKEDATA
              Copy the word data to the address addr in the tracee's memory.  As for PTRACE_PEEKTEXT and PTRACE_PEEKDATA, these two requests are currently equivalent.

       PTRACE_POKEUSER
              Copy the word data to offset addr in the tracee's USER area.  As for PTRACE_PEEKUSER, the offset must typically be word-aligned.  In order to maintain the integrity of the kernel, some modifications to the USER area are
              disallowed.

       PTRACE_GETREGS, PTRACE_GETFPREGS
              Copy  the tracee's general-purpose or floating-point registers, respectively, to the address data in the tracer.  See <sys/user.h> for information on the format of this data.  (addr is ignored.)  Note that SPARC systems
              have the meaning of data and addr reversed; that is, data is ignored and the registers are copied to the address addr.  PTRACE_GETREGS and PTRACE_GETFPREGS are not present on all architectures.

       PTRACE_GETREGSET (since Linux 2.6.34)
              Read the tracee's registers.  addr specifies, in an architecture-dependent way, the type of registers to be read.  NT_PRSTATUS (with numerical value 1) usually results in reading of general-purpose  registers.   If  the
              CPU has, for example, floating-point and/or vector registers, they can be retrieved by setting addr to the corresponding NT_foo constant.  data points to a struct iovec, which describes the destination buffer's location
              and length.  On return, the kernel modifies iov.len to indicate the actual number of bytes returned.

       PTRACE_SETREGS, PTRACE_SETFPREGS
              Modify the tracee's general-purpose or floating-point registers, respectively, from the address data in the tracer.  As for PTRACE_POKEUSER, some general-purpose register  modifications  may  be  disallowed.   (addr  is
              ignored.)   Note  that  SPARC  systems have the meaning of data and addr reversed; that is, data is ignored and the registers are copied from the address addr.  PTRACE_SETREGS and PTRACE_SETFPREGS are not present on all
              architectures.

       PTRACE_SETREGSET (since Linux 2.6.34)
              Modify the tracee's registers.  The meaning of addr and data is analogous to PTRACE_GETREGSET.

       PTRACE_GETSIGINFO (since Linux 2.3.99-pre6)
              Retrieve information about the signal that caused the stop.  Copy a siginfo_t structure (see sigaction(2)) from the tracee to the address data in the tracer.  (addr is ignored.)

       PTRACE_SETSIGINFO (since Linux 2.3.99-pre6)
              Set signal information: copy a siginfo_t structure from the address data in the tracer to the tracee.  This will affect only signals that would normally be delivered to the tracee and were caught by the tracer.  It  may
              be difficult to tell these normal signals from synthetic signals generated by ptrace() itself.  (addr is ignored.)

       PTRACE_PEEKSIGINFO (since Linux 3.10)
              Retrieve  siginfo_t structures without removing signals from a queue.  addr points to a ptrace_peeksiginfo_args structure that specifies the ordinal position from which copying of signals should start, and the number of
              signals to copy.  siginfo_t structures are copied into the buffer pointed to by data.  The return value contains the number of copied signals (zero indicates that there is no signal corresponding to the specified  ordi‐
              nal position).  Within the returned siginfo structures, the si_code field includes information (__SI_CHLD, __SI_FAULT, etc.) that are not otherwise exposed to user space.

                 struct ptrace_peeksiginfo_args {
                     u64 off;    /* Ordinal position in queue at which
                                    to start copying signals */
                     u32 flags;  /* PTRACE_PEEKSIGINFO_SHARED or 0 */
                     s32 nr;     /* Number of signals to copy */
                 };

                 Currently, there is only one flag, PTRACE_PEEKSIGINFO_SHARED, for dumping signals from the process-wide signal queue.  If this flag is not set, signals are read from the per-thread queue of the specified thread.

       PTRACE_GETSIGMASK (since Linux 3.11)
              Place a copy of the mask of blocked signals (see sigprocmask(2)) in the buffer pointed to by data, which should be a pointer to a buffer of type sigset_t.  The addr argument contains the size of the buffer pointed to by
              data (i.e., sizeof(sigset_t)).

       PTRACE_SETSIGMASK (since Linux 3.11)
              Change the mask of blocked signals (see sigprocmask(2)) to the value specified in the buffer pointed to by data, which should be a pointer to a buffer of type sigset_t.  The addr argument contains the size of the buffer
              pointed to by data (i.e., sizeof(sigset_t)).

       PTRACE_SETOPTIONS (since Linux 2.4.6; see BUGS for caveats)
              Set ptrace options from data.  (addr is ignored.)  data is interpreted as a bit mask of options, which are specified by the following flags:

              PTRACE_O_EXITKILL (since Linux 3.8)
                     If a tracer sets this flag, a SIGKILL signal will be sent to every tracee if the tracer exits.  This option is useful for ptrace jailers that want to ensure that tracees can never escape the tracer's control.

              PTRACE_O_TRACECLONE (since Linux 2.5.46)
                     Stop  the  tracee  at  the  next clone(2) and automatically start tracing the newly cloned process, which will start with a SIGSTOP, or PTRACE_EVENT_STOP if PTRACE_SEIZE was used.  A waitpid(2) by the tracer will
                     return a status value such that

                       status>>8 == (SIGTRAP | (PTRACE_EVENT_CLONE<<8))

                     The PID of the new process can be retrieved with PTRACE_GETEVENTMSG.

                     This option may not catch clone(2) calls in all cases.  If the tracee calls clone(2) with the CLONE_VFORK flag, PTRACE_EVENT_VFORK will be delivered instead if PTRACE_O_TRACEVFORK is set; otherwise if the  tracee
                     calls clone(2) with the exit signal set to SIGCHLD, PTRACE_EVENT_FORK will be delivered if PTRACE_O_TRACEFORK is set.

              PTRACE_O_TRACEEXEC (since Linux 2.5.46)
                     Stop the tracee at the next execve(2).  A waitpid(2) by the tracer will return a status value such that

                       status>>8 == (SIGTRAP | (PTRACE_EVENT_EXEC<<8))

                     If the execing thread is not a thread group leader, the thread ID is reset to thread group leader's ID before this stop.  Since Linux 3.0, the former thread ID can be retrieved with PTRACE_GETEVENTMSG.

              PTRACE_O_TRACEEXIT (since Linux 2.5.60)
                     Stop the tracee at exit.  A waitpid(2) by the tracer will return a status value such that

                       status>>8 == (SIGTRAP | (PTRACE_EVENT_EXIT<<8))

                     The tracee's exit status can be retrieved with PTRACE_GETEVENTMSG.

                     The  tracee is stopped early during process exit, when registers are still available, allowing the tracer to see where the exit occurred, whereas the normal exit notification is done after the process is finished
                     exiting.  Even though context is available, the tracer cannot prevent the exit from happening at this point.

              PTRACE_O_TRACEFORK (since Linux 2.5.46)
                     Stop the tracee at the next fork(2) and automatically start tracing the newly forked process, which will start with a SIGSTOP, or PTRACE_EVENT_STOP if PTRACE_SEIZE was used.   A  waitpid(2)  by  the  tracer  will
                     return a status value such that

                       status>>8 == (SIGTRAP | (PTRACE_EVENT_FORK<<8))

                     The PID of the new process can be retrieved with PTRACE_GETEVENTMSG.

              PTRACE_O_TRACESYSGOOD (since Linux 2.4.6)
                     When  delivering  system  call  traps,  set  bit  7  in  the  signal  number  (i.e.,  deliver  SIGTRAP|0x80).   This  makes  it  easy for the tracer to distinguish normal traps from those caused by a system call.
                     (PTRACE_O_TRACESYSGOOD may not work on all architectures.)

              PTRACE_O_TRACEVFORK (since Linux 2.5.46)
                     Stop the tracee at the next vfork(2) and automatically start tracing the newly vforked process, which will start with a SIGSTOP, or PTRACE_EVENT_STOP if PTRACE_SEIZE was used.  A waitpid(2)  by  the  tracer  will
                     return a status value such that

                       status>>8 == (SIGTRAP | (PTRACE_EVENT_VFORK<<8))

                     The PID of the new process can be retrieved with PTRACE_GETEVENTMSG.

              PTRACE_O_TRACEVFORKDONE (since Linux 2.5.60)
                     Stop the tracee at the completion of the next vfork(2).  A waitpid(2) by the tracer will return a status value such that

                       status>>8 == (SIGTRAP | (PTRACE_EVENT_VFORK_DONE<<8))

                     The PID of the new process can (since Linux 2.6.18) be retrieved with PTRACE_GETEVENTMSG.

              PTRACE_O_TRACESECCOMP (since Linux 3.5)
                     Stop the tracee when a seccomp(2) SECCOMP_RET_TRACE rule is triggered.  A waitpid(2) by the tracer will return a status value such that

                       status>>8 == (SIGTRAP | (PTRACE_EVENT_SECCOMP<<8))

                     While  this  triggers  a  PTRACE_EVENT  stop,  it  is  similar to a syscall-enter-stop, in that the tracee has not yet entered the syscall that seccomp triggered on.  The seccomp event message data (from the SEC‐
                     COMP_RET_DATA portion of the seccomp filter rule) can be retrieved with PTRACE_GETEVENTMSG.

              PTRACE_O_SUSPEND_SECCOMP (since Linux 4.3)
                     Suspend the tracee's seccomp protections.  This applies regardless of mode, and can be used when the tracee has not yet installed seccomp filters.  That is, a valid use case is to suspend a tracee's seccomp  pro‐
                     tections  before  they  are  installed  by  the  tracee,  let  the  tracee  install the filters, and then clear this flag when the filters should be resumed.  Setting this option requires that the tracer have the
                     CAP_SYS_ADMIN capability, not have any seccomp protections installed, and not have PTRACE_O_SUSPEND_SECCOMP set on itself.

       PTRACE_GETEVENTMSG (since Linux 2.5.46)
              Retrieve a message (as an unsigned long) about the ptrace event that just happened, placing it at the address data in the tracer.  For PTRACE_EVENT_EXIT,  this  is  the  tracee's  exit  status.   For  PTRACE_EVENT_FORK,
              PTRACE_EVENT_VFORK,  PTRACE_EVENT_VFORK_DONE,  and  PTRACE_EVENT_CLONE, this is the PID of the new process.  For PTRACE_EVENT_SECCOMP, this is the seccomp(2) filter's SECCOMP_RET_DATA associated with the triggered rule.
              (addr is ignored.)

       PTRACE_CONT
              Restart the stopped tracee process.  If data is nonzero, it is interpreted as the number of a signal to be delivered to the tracee; otherwise, no signal is delivered.  Thus, for example, the tracer can control whether a
              signal sent to the tracee is delivered or not.  (addr is ignored.)

       PTRACE_SYSCALL, PTRACE_SINGLESTEP
              Restart  the stopped tracee as for PTRACE_CONT, but arrange for the tracee to be stopped at the next entry to or exit from a system call, or after execution of a single instruction, respectively.  (The tracee will also,
              as usual, be stopped upon receipt of a signal.)  From the tracer's perspective, the tracee will appear to have been stopped by receipt of a SIGTRAP.  So, for PTRACE_SYSCALL, for example, the idea is to inspect the argu‐
              ments to the system call at the first stop, then do another PTRACE_SYSCALL and inspect the return value of the system call at the second stop.  The data argument is treated as for PTRACE_CONT.  (addr is ignored.)

       PTRACE_SYSEMU, PTRACE_SYSEMU_SINGLESTEP (since Linux 2.6.14)
              For PTRACE_SYSEMU, continue and stop on entry to the next system call, which will not be executed.  For PTRACE_SYSEMU_SINGLESTEP, do the same but also singlestep if not a system call.  This call is used by programs like
              User Mode Linux that want to emulate all the tracee's system calls.  The data argument is treated as for PTRACE_CONT.  The addr argument is ignored.  These requests are currently supported only on x86.

       PTRACE_LISTEN (since Linux 3.4)
              Restart the stopped tracee, but prevent it from executing.  The resulting state of the tracee is similar to a process which has been stopped by a SIGSTOP (or other stopping signal).  See the "group-stop" subsection  for
              additional information.  PTRACE_LISTEN works only on tracees attached by PTRACE_SEIZE.

       PTRACE_KILL
              Send the tracee a SIGKILL to terminate it.  (addr and data are ignored.)

              This operation is deprecated; do not use it!  Instead, send a SIGKILL directly using kill(2) or tgkill(2).  The problem with PTRACE_KILL is that it requires the tracee to be in signal-delivery-stop, otherwise it may not
              work (i.e., may complete successfully but won't kill the tracee).  By contrast, sending a SIGKILL directly has no such limitation.

       PTRACE_INTERRUPT (since Linux 3.4)
              Stop a tracee.  If the tracee is running or sleeping in kernel space and PTRACE_SYSCALL is in effect, the system call is interrupted and syscall-exit-stop is reported.  (The interrupted system call is restarted when the
              tracee  is  restarted.)  If the tracee was already stopped by a signal and PTRACE_LISTEN was sent to it, the tracee stops with PTRACE_EVENT_STOP and WSTOPSIG(status) returns the stop signal.  If any other ptrace-stop is
              generated at the same time (for example, if a signal is sent to the tracee), this ptrace-stop happens.  If none of the above applies (for example, if the tracee is running in user space), it stops with PTRACE_EVENT_STOP
              with WSTOPSIG(status) == SIGTRAP.  PTRACE_INTERRUPT only works on tracees attached by PTRACE_SEIZE.

       PTRACE_ATTACH
              Attach  to  the  process  specified  in pid, making it a tracee of the calling process.  The tracee is sent a SIGSTOP, but will not necessarily have stopped by the completion of this call; use waitpid(2) to wait for the
              tracee to stop.  See the "Attaching and detaching" subsection for additional information.  (addr and data are ignored.)

              Processes that are not dumpable (see the description of PR_SET_DUMPABLE in prctl(2)) can  not  be  attached  via ptrace(2) PTRACE_ATTACH.

       PTRACE_SEIZE (since Linux 3.4)
              Attach to the process specified in pid, making it a tracee of the calling process.  Unlike PTRACE_ATTACH, PTRACE_SEIZE does not stop the process.  Group-stops  are  reported  as  PTRACE_EVENT_STOP  and  WSTOPSIG(status)
              returns  the  stop  signal.  Automatically attached children stop with PTRACE_EVENT_STOP and WSTOPSIG(status) returns SIGTRAP instead of having SIGSTOP signal delivered to them.  execve(2) does not deliver an extra SIG‐
              TRAP.  Only a PTRACE_SEIZEd process can accept PTRACE_INTERRUPT and PTRACE_LISTEN commands.  The "seized" behavior just described is inherited by  children  that  are  automatically  attached  using  PTRACE_O_TRACEFORK,
              PTRACE_O_TRACEVFORK, and PTRACE_O_TRACECLONE.  addr must be zero.  data contains a bit mask of ptrace options to activate immediately.

       PTRACE_DETACH
              Restart the stopped tracee as for PTRACE_CONT, but first detach from it.  Under Linux, a tracee can be detached in this way regardless of which method was used to initiate tracing.  (addr is ignored.)

   Death under ptrace
       When  a  (possibly  multithreaded)  process receives a killing signal (one whose disposition is set to SIG_DFL and whose default action is to kill the process), all threads exit.  Tracees report their death to their tracer(s).
       Notification of this event is delivered via waitpid(2).

       Note that the killing signal will first cause signal-delivery-stop (on one tracee only), and only after it is injected by the tracer (or after it was dispatched to a thread which isn't traced), will death from the signal  hap‐
       pen on all tracees within a multithreaded process.  (The term "signal-delivery-stop" is explained below.)

       SIGKILL does not generate signal-delivery-stop and therefore the tracer can't suppress it.  SIGKILL kills even within system calls (syscall-exit-stop is not generated prior to death by SIGKILL).  The net effect is that SIGKILL
       always kills the process (all its threads), even if some threads of the process are ptraced.

       When the tracee calls _exit(2), it reports its death to its tracer.  Other threads are not affected.

       When any thread executes exit_group(2), every tracee in its thread group reports its death to its tracer.

       If the PTRACE_O_TRACEEXIT option is on, PTRACE_EVENT_EXIT will happen before actual death.  This applies to exits via exit(2), exit_group(2), and signal deaths (except SIGKILL, depending on the kernel version; see BUGS below),
       and when threads are torn down on execve(2) in a multithreaded process.

       The  tracer cannot assume that the ptrace-stopped tracee exists.  There are many scenarios when the tracee may die while stopped (such as SIGKILL).  Therefore, the tracer must be prepared to handle an ESRCH error on any ptrace
       operation.  Unfortunately, the same error is returned if the tracee exists but is not ptrace-stopped (for commands which require a stopped tracee), or if it is not traced by the process  which  issued  the  ptrace  call.   The
       tracer needs to keep track of the stopped/running state of the tracee, and interpret ESRCH as "tracee died unexpectedly" only if it knows that the tracee has been observed to enter ptrace-stop.  Note that there is no guarantee
       that waitpid(WNOHANG) will reliably report the tracee's death status if a ptrace operation returned ESRCH.  waitpid(WNOHANG) may return 0 instead.  In other words, the tracee may be "not yet fully dead", but  already  refusing
       ptrace requests.

       The  tracer  can't  assume that the tracee always ends its life by reporting WIFEXITED(status) or WIFSIGNALED(status); there are cases where this does not occur.  For example, if a thread other than thread group leader does an
       execve(2), it disappears; its PID will never be seen again, and any subsequent ptrace stops will be reported under the thread group leader's PID.

   Stopped states
       A tracee can be in two states: running or stopped.  For the purposes of ptrace, a tracee which is blocked in a system call (such as read(2), pause(2), etc.)  is nevertheless considered to be running,  even  if  the  tracee  is
       blocked  for  a long time.  The state of the tracee after PTRACE_LISTEN is somewhat of a gray area: it is not in any ptrace-stop (ptrace commands won't work on it, and it will deliver waitpid(2) notifications), but it also may
       be considered "stopped" because it is not executing instructions (is not scheduled), and if it was in group-stop before PTRACE_LISTEN, it will not respond to signals until SIGCONT is received.

       There are many kinds of states when the tracee is stopped, and in ptrace discussions they are often conflated.  Therefore, it is important to use precise terms.

       In this manual page, any stopped state in which the tracee is ready to accept ptrace commands from the tracer is called ptrace-stop.  Ptrace-stops can be further subdivided into signal-delivery-stop, group-stop,  syscall-stop,
       and so on.  These stopped states are described in detail below.

       When the running tracee enters ptrace-stop, it notifies its tracer using waitpid(2) (or one of the other "wait" system calls).  Most of this manual page assumes that the tracer waits with:

           pid = waitpid(pid_or_minus_1, &status, __WALL);

       Ptrace-stopped tracees are reported as returns with pid greater than 0 and WIFSTOPPED(status) true.

       The __WALL flag does not include the WSTOPPED and WEXITED flags, but implies their functionality.

       Setting the WCONTINUED flag when calling waitpid(2) is not recommended: the "continued" state is per-process and consuming it can confuse the real parent of the tracee.

       Use of the WNOHANG flag may cause waitpid(2) to return 0 ("no wait results available yet") even if the tracer knows there should be a notification.  Example:

           errno = 0;
           ptrace(PTRACE_CONT, pid, 0L, 0L);
           if (errno == ESRCH) {
               /* tracee is dead */
               r = waitpid(tracee, &status, __WALL | WNOHANG);
               /* r can still be 0 here! */
           }

       The  following kinds of ptrace-stops exist: signal-delivery-stops, group-stops, PTRACE_EVENT stops, syscall-stops.  They all are reported by waitpid(2) with WIFSTOPPED(status) true.  They may be differentiated by examining the
       value status>>8, and if there is ambiguity in that value, by querying PTRACE_GETSIGINFO.  (Note: the WSTOPSIG(status) macro can't be used to perform this examination, because it returns the value (status>>8) & 0xff.)

   Signal-delivery-stop
       When a (possibly multithreaded) process receives any signal except SIGKILL, the kernel selects an arbitrary thread which handles the signal.  (If the signal is generated with tgkill(2), the  target  thread  can  be  explicitly
       selected  by  the caller.)  If the selected thread is traced, it enters signal-delivery-stop.  At this point, the signal is not yet delivered to the process, and can be suppressed by the tracer.  If the tracer doesn't suppress
       the signal, it passes the signal to the tracee in the next ptrace restart request.  This second step of signal delivery is called signal injection in this manual page.  Note that if the signal is blocked,  signal-delivery-stop
       doesn't happen until the signal is unblocked, with the usual exception that SIGSTOP can't be blocked.

       Signal-delivery-stop is observed by the tracer as waitpid(2) returning with WIFSTOPPED(status) true, with the signal returned by WSTOPSIG(status).  If the signal is SIGTRAP, this may be a different kind of ptrace-stop; see the
       "Syscall-stops" and "execve" sections below for details.  If WSTOPSIG(status) returns a stopping signal, this may be a group-stop; see below.

   Signal injection and suppression
       After signal-delivery-stop is observed by the tracer, the tracer should restart the tracee with the call

           ptrace(PTRACE_restart, pid, 0, sig)

       where PTRACE_restart is one of the restarting ptrace requests.  If sig is 0, then a signal is not delivered.  Otherwise, the signal sig is delivered.  This operation is called signal injection in this manual page,  to  distin‐
       guish it from signal-delivery-stop.

       The sig value may be different from the WSTOPSIG(status) value: the tracer can cause a different signal to be injected.

       Note  that  a suppressed signal still causes system calls to return prematurely.  In this case, system calls will be restarted: the tracer will observe the tracee to reexecute the interrupted system call (or restart_syscall(2)
       system call for a few system calls which use a different mechanism for restarting) if the tracer uses PTRACE_SYSCALL.  Even system calls (such as poll(2)) which are not restartable after signal are restarted  after  signal  is
       suppressed; however, kernel bugs exist which cause some system calls to fail with EINTR even though no observable signal is injected to the tracee.

       Restarting  ptrace commands issued in ptrace-stops other than signal-delivery-stop are not guaranteed to inject a signal, even if sig is nonzero.  No error is reported; a nonzero sig may simply be ignored.  Ptrace users should
       not try to "create a new signal" this way: use tgkill(2) instead.

       The fact that signal injection requests may be ignored when restarting the tracee after ptrace stops that are not signal-delivery-stops is a cause of confusion among ptrace users.  One  typical  scenario  is  that  the  tracer
       observes group-stop, mistakes it for signal-delivery-stop, restarts the tracee with

           ptrace(PTRACE_restart, pid, 0, stopsig)

       with the intention of injecting stopsig, but stopsig gets ignored and the tracee continues to run.

       The SIGCONT signal has a side effect of waking up (all threads of) a group-stopped process.  This side effect happens before signal-delivery-stop.  The tracer can't suppress this side effect (it can only suppress signal injec‐
       tion, which only causes the SIGCONT handler to not be executed in the tracee, if such a handler is installed).  In fact, waking up from group-stop may be followed by signal-delivery-stop for signal(s) other  than  SIGCONT,  if
       they were pending when SIGCONT was delivered.  In other words, SIGCONT may be not the first signal observed by the tracee after it was sent.

       Stopping signals cause (all threads of) a process to enter group-stop.  This side effect happens after signal injection, and therefore can be suppressed by the tracer.

       In Linux 2.4 and earlier, the SIGSTOP signal can't be injected.

       PTRACE_GETSIGINFO can be used to retrieve a siginfo_t structure which corresponds to the delivered signal.  PTRACE_SETSIGINFO may be used to modify it.  If PTRACE_SETSIGINFO has been used to alter siginfo_t, the si_signo field
       and the sig parameter in the restarting command must match, otherwise the result is undefined.

   Group-stop
       When a (possibly multithreaded) process receives a stopping signal, all threads stop.  If some threads are traced, they enter a group-stop.  Note that the stopping signal will first cause signal-delivery-stop  (on  one  tracee
       only),  and only after it is injected by the tracer (or after it was dispatched to a thread which isn't traced), will group-stop be initiated on all tracees within the multithreaded process.  As usual, every tracee reports its
       group-stop separately to the corresponding tracer.

       Group-stop is observed by the tracer as waitpid(2) returning with WIFSTOPPED(status) true, with the stopping signal available via WSTOPSIG(status).  The same result is returned by some other classes of ptrace-stops,  therefore
       the recommended practice is to perform the call

           ptrace(PTRACE_GETSIGINFO, pid, 0, &siginfo)


source/ptrace.d
/**
  * ptrace estas linuksa sistemo, ke ebligas vin vidi en aliajn procedojn.
  *
  * Ĉi tiu dokumentaro estas en la angan, ĉar ni ĵus kopiis la linuksan dokumentaron.
  */
module initd.ptrace;

import core.stdc.config : c_long;
import core.sys.posix.sys.types : pid_t;

void afiksu(int pido)
{
    import core.stdc.errno : errno;
    import core.stdc.string : strerror;

    auto rezulto = ptrace(PtraceRequest.ATTACH, cast(pid_t) pido, null, null);
    if (rezulto == -1)
    {
        import std.string : fromStringz;
        import std.format : format;

        auto eraro = errno;
        auto ĉeno = strerror(eraro);
        auto eraroĈeno = ĉeno.fromStringz.idup;
        free(ĉeno);
        throw new PtraceException(
                "malplenumis afiksi al pido %s: %s".format(
                    pido, eraroĈeno));
    }
}

class PtraceException : Exception
{
    this(string mensaĝo) { super(mensaĝo); }
}

extern(C) c_long ptrace(PtraceRequest request, pid_t pid, void* address, void* data);

enum PtraceRequest
{
    /** Indicate that the process making this request should be traced.
      All signals received by this process can be intercepted by its
      parent, and its parent can use the other `ptrace' requests.  */
    TRACEME = 0,

    /** Return the word in the process's text space at address ADDR.  */
    PEEKTEXT = 1,

    /** Return the word in the process's data space at address ADDR.  */
    PEEKDATA = 2,

    /** Return the word in the process's user area at offset ADDR.  */
    PEEKUSER = 3,

    /** Write the word DATA into the process's text space at address ADDR.  */
    POKETEXT = 4,

    /** Write the word DATA into the process's data space at address ADDR.  */
    POKEDATA = 5,

    /** Write the word DATA into the process's user area at offset ADDR.  */
    POKEUSER = 6,

    /** Continue the process.  */
    CONT = 7,

    /** Kill the process.  */
    KILL = 8,

    /** Single step the process.
      This is not supported on all machines.  */
    SINGLESTEP = 9,

    /** Get all general purpose registers used by a processes.
      This is not supported on all machines.  */
    GETREGS = 12,

    /** Set all general purpose registers used by a processes.
      This is not supported on all machines.  */
    SETREGS = 13,

    /** Get all floating point registers used by a processes.
      This is not supported on all machines.  */
    GETFPREGS = 14,

    /** Set all floating point registers used by a processes.
      This is not supported on all machines.  */
    SETFPREGS = 15,

    /** Attach to a process that is already running. */
    ATTACH = 16,

    /** Detach from a process attached to with ATTACH.  */
    DETACH = 17,

    /** Get all extended floating point registers used by a processes.
      This is not supported on all machines.  */
    GETFPXREGS = 18,

    /** Set all extended floating point registers used by a processes.
      This is not supported on all machines.  */
    SETFPXREGS = 19,

    /** Continue and stop at the next (return from) syscall.  */
    SYSCALL = 24,

    /** Set ptrace filter options.  */
    SETOPTIONS = 0x4200,

    /** Get last ptrace message.  */
    GETEVENTMSG = 0x4201,

    /** Get siginfo for process.  */
    GETSIGINFO = 0x4202,

    /** Set new siginfo for process.  */
    SETSIGINFO = 0x4203,

    /** Get register content.  */
    GETREGSET = 0x4204,

    /** Set register content.  */
    SETREGSET = 0x4205,

    /** Like ATTACH, but do not force tracee to trap and do not affect
      signal or group stop state.  */
    SEIZE = 0x4206,

    /** Trap seized tracee.  */
    INTERRUPT = 0x4207,

    /** Wait for next group event.  */
    LISTEN = 0x4208,

    PEEKSIGINFO = 0x4209,

    GETSIGMASK = 0x420a,

    SETSIGMASK = 0x420b,

    SECCOMP_GET_FILTER = 0x420c
}


/** Flag for LISTEN.  */
enum Flags
{
    SEIZE_DEVEL = 0x80000000
}

/** Options set using SETOPTIONS.  */
enum Options
{
    TRACESYSGOOD = 0x00000001,
    TRACEFORK = 0x00000002,
    TRACEVFORK   = 0x00000004,
    TRACECLONE = 0x00000008,
    TRACEEXEC = 0x00000010,
    TRACEVFORKDONE = 0x00000020,
    TRACEEXIT = 0x00000040,
    TRACESECCOMP = 0x00000080,
    EXITKILL = 0x00100000,
    SUSPEND_SECCOMP = 0x00200000,
    MASK  = 0x003000ff
}

/** Wait extended result codes for the above trace options.  */
enum __ptrace_eventcodes
{
    EVENT_FORK = 1,
    EVENT_VFORK = 2,
    EVENT_CLONE = 3,
    EVENT_EXEC = 4,
    EVENT_VFORK_DONE = 5,
    EVENT_EXIT = 6,
    EVENT_SECCOMP  = 7
}

/** Arguments for PEEKSIGINFO.  */
struct PeekSigInfoArgs
{
    ulong off; /** From which siginfo to start.  */
    PeekSigInfoFlags flags; /** Flags for peeksiginfo.  */
    int nr;  /** How many siginfos to take.  */
}

enum PeekSigInfoFlags
{
    /** Read signals from a shared (process wide) queue.  */
    PEEKSIGINFO_SHARED = (1 << 0)
}
